common/lib/xmodule/xmodule/modulestore/courseware_index.py

```diff
@@ -0,0 +1,118 @@
+""" Code to allow module store to interface with courseware index """
+from __future__ import absolute_import
+
+import logging
+
+from opaque_keys.edx.locator import CourseLocator
+from search.search_engine_base import SearchEngine
+
+from . import ModuleStoreEnum
+from .exceptions import ItemNotFoundError
+
+# Use default index and document names for now
+INDEX_NAME = "courseware_index"
+DOCUMENT_TYPE = "courseware_content"
+
+log = logging.getLogger('edx.modulestore')
+
+
+class IndexWriteError(Exception):
+
+    """ Raised to indicate that indexing of particular key failed """
+    pass
+
+
+class ModuleStoreCoursewareIndexMixin(object):
+
+    """
+    Mixin class to enable indexing for courseware search from different modulestores
+    """
+
+    def do_index(self, location, delete=False):
+        """
+        Main routine to index (for purposes of searching) from given location and other stuff on down
+        """
+        error = []
+        # TODO - inline for now, need to move this out to a celery task
+        searcher = SearchEngine.get_search_engine(INDEX_NAME)
+        if not searcher:
+            return
+
+        course_key = location if isinstance(location, CourseLocator) else location.course_key
+        location_info = {
+            "course": unicode(course_key),
+        }
+
+        def _fetch_item(item_location):
+            """ Fetch the item from the modulestore location, log if not found, but continue """
+            try:
+                if isinstance(item_location, CourseLocator):
+                    item = self.get_course(item_location)
+                else:
+                    item = self.get_item(item_location, revision=ModuleStoreEnum.RevisionOption.published_only)
+            except ItemNotFoundError:
```

Comment:
We should also chat later about whether we should be passing in depth=... parameters on the Mongo fetches to cut down on the # round-trips. Not a blocker now.